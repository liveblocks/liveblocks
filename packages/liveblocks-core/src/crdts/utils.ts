import type { Json } from "../lib/Json";
import type { PlainLson } from "../types/PlainLson";
import { LiveList } from "./LiveList";
import { LiveMap } from "./LiveMap";
import { LiveObject } from "./LiveObject";
import type { Lson, LsonObject } from "./Lson";

/**
 * Helper type to convert any valid Lson type to the equivalent Json type.
 *
 * Examples:
 *
 *   ToImmutable<42>                         // 42
 *   ToImmutable<'hi'>                       // 'hi'
 *   ToImmutable<number>                     // number
 *   ToImmutable<string>                     // string
 *   ToImmutable<string | LiveList<number>>  // string | readonly number[]
 *   ToImmutable<LiveMap<string, LiveList<number>>>
 *                                           // ReadonlyMap<string, readonly number[]>
 *   ToImmutable<LiveObject<{ a: number, b: LiveList<string>, c?: number }>>
 *                                           // { readonly a: null, readonly b: readonly string[], readonly c?: number }
 *
 */
// prettier-ignore
export type ToImmutable<L extends Lson | LsonObject> =
  // A LiveList serializes to an equivalent JSON array
  L extends LiveList<infer I> ? readonly ToImmutable<I>[] :

  // A LiveObject serializes to an equivalent JSON object
  L extends LiveObject<infer O> ? ToImmutable<O> :

  // A LiveMap serializes to a JSON object with string-V pairs
  L extends LiveMap<infer K, infer V> ? ReadonlyMap<K, ToImmutable<V>> :

  // Any LsonObject recursively becomes a JsonObject
  L extends LsonObject ?
    { readonly [K in keyof L]: ToImmutable<Exclude<L[K], undefined>>
                                 | (undefined extends L[K] ? undefined : never) } :

  // Any Json value already is a legal Json value
  L extends Json ? L :

  // Otherwise, this is not possible
  never;

/**
 * Returns PlainLson for a given Json or LiveStructure, suitable for calling the storage init api
 */
export function toPlainLson(lson: Lson): PlainLson {
  if (lson instanceof LiveObject) {
    return {
      liveblocksType: "LiveObject",
      data: Object.fromEntries(
        Object.entries(lson.toObject()).flatMap(([key, value]) =>
          value !== undefined ? [[key, toPlainLson(value)]] : []
        )
      ),
    };
  } else if (lson instanceof LiveMap) {
    return {
      liveblocksType: "LiveMap",
      data: Object.fromEntries(
        [...lson].map(([key, value]) => [key, toPlainLson(value)])
      ),
    };
  } else if (lson instanceof LiveList) {
    return {
      liveblocksType: "LiveList",
      data: [...lson].map((item) => toPlainLson(item)),
    };
  } else {
    return lson;
  }
}
