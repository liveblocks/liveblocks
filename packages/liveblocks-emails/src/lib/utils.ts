export const isString = (value: unknown): value is string => {
  return typeof value === "string";
};

export const isMentionNodeAttributeId = (
  value: unknown
): value is `in_${string}` => {
  return isString(value) && value.startsWith("in_");
};
