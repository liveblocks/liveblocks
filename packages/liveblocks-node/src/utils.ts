export function assertNonEmpty(
  value: unknown,
  field: string
): asserts value is string {
  if (typeof value !== "string" || value.length === 0) {
    throw new Error(
      `Invalid value for field "${field}". Please provide a non-empty string. For more information: https://liveblocks.io/docs/api-reference/liveblocks-node#authorize`
    );
  }
}
