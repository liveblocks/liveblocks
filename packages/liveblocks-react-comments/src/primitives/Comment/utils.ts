import type { CommentBodyElement, CommentBodyMention } from "@liveblocks/core";

export function isCommentBodyMention(
  element: CommentBodyElement
): element is CommentBodyMention {
  return "type" in element && element.type === "mention";
}
