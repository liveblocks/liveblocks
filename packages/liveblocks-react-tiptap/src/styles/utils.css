@mixin invisible-selection {
  &::selection,
  *::selection {
    background: transparent;
  }
}

@mixin invisible-scrollbar {
  & {
    -ms-overflow-style: none;
    scrollbar-width: none;
  }

  &::-webkit-scrollbar {
    display: none;
  }
}

@mixin truncate {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}
