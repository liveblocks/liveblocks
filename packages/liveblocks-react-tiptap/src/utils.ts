import type { ClientRectObject } from "@floating-ui/react-dom";
import type { ContextualPromptContext } from "@liveblocks/core";
import type { Editor, Range } from "@tiptap/core";
import type { Node as ProseMirrorNode } from "@tiptap/pm/model";
import { Fragment } from "@tiptap/pm/model";
import {
  type EditorState,
  type Selection,
  TextSelection,
} from "@tiptap/pm/state";
import {
  getRelativeSelection,
  relativePositionToAbsolutePosition,
  ySyncPluginKey,
} from "y-prosemirror";
import type { RelativePosition } from "yjs";

import type { YSyncPluginState } from "./types";
import { LIVEBLOCKS_MENTION_TYPE } from "./types";

const CONTEXT_TRUNCATION = "[â€¦]";
const CONTEXT_BLOCK_SEPARATOR = "\n";

export const getRelativeSelectionFromState = (state: EditorState) => {
  const pluginState = ySyncPluginKey.getState(state) as YSyncPluginState;
  if (!pluginState) return null;
  return getRelativeSelection(pluginState.binding, state);
};

export const getRangeFromRelativeSelections = (
  pos: { anchor: RelativePosition; head: RelativePosition },
  state: EditorState
) => {
  const pluginState = ySyncPluginKey.getState(state) as YSyncPluginState;
  if (!pluginState || !pluginState.binding) return { from: 0, to: 0 };
  const { doc, type, mapping } = pluginState.binding;
  const anchor =
    relativePositionToAbsolutePosition(doc, type, pos.anchor, mapping) ?? 0;
  const head =
    relativePositionToAbsolutePosition(doc, type, pos.head, mapping) ?? 0;

  const from = anchor > head ? head : anchor;
  const to = anchor > head ? anchor : head;
  return { from, to };
};

export const getRectFromCoords = (coords: {
  top: number;
  left: number;
  right: number;
  bottom: number;
}): ClientRectObject => {
  return {
    ...coords,
    x: coords.left,
    y: coords.top,
    width: coords.right - coords.left,
    height: coords.bottom - coords.top,
  };
};

export const getMentionsFromNode = (
  node: ProseMirrorNode,
  range: Range
): { notificationId: string; userId: string }[] => {
  const result: { notificationId: string; userId: string }[] = [];
  node.nodesBetween(range.from, range.to, (child) => {
    if (child.type.name === LIVEBLOCKS_MENTION_TYPE) {
      const mention = child.attrs as { id?: string; notificationId?: string };
      if (mention.id && mention.notificationId) {
        result.push({
          notificationId: mention.notificationId,
          userId: mention.id,
        });
      }
    }
  });
  return result;
};

// How to modify data in transformPasted, inspired by: https://discuss.prosemirror.net/t/modify-specific-node-on-copy-and-paste-in-clipboard/4901/4
export const mapFragment = (
  fragment: Fragment,
  callback: (
    node: ProseMirrorNode
  ) => ProseMirrorNode | ProseMirrorNode[] | Fragment | null
): Fragment => {
  const content: ProseMirrorNode[] = [];
  fragment.forEach((node) => {
    if (node.content.childCount > 0) {
      content.push(
        node.type.create(node.attrs, mapFragment(node.content, callback))
      );
      return;
    }
    content.push(callback(node) as ProseMirrorNode);
  });

  return Fragment.from(content);
};

export function getDomRangeFromSelection(editor: Editor, selection: Selection) {
  if (selection.from === selection.to) {
    const { parent, parentOffset } = selection.$from;

    // If the selection is collapsed and in an empty block node or at the end
    // of a text node, extend it to the entire node
    if (
      (parent.isBlock && parent.content.size === 0) ||
      (parent.isTextblock && parentOffset === parent.content.size)
    ) {
      selection = TextSelection.create(
        editor.state.doc,
        selection.$from.before(),
        selection.$from.after()
      );
    }
  }

  const from = editor.view.domAtPos(selection.from);
  const to = editor.view.domAtPos(selection.to);

  const domRange = document.createRange();
  domRange.setStart(from.node, from.offset);
  domRange.setEnd(to.node, to.offset);

  return domRange;
}

export function compareSelections(
  a: Selection | null | undefined,
  b: Selection | null | undefined
) {
  if (!a || !b) {
    return false;
  }

  return a.eq(b);
}

export function getContextualPromptContext(
  editor: Editor,
  maxLength = 10_000
): ContextualPromptContext {
  const { selection, doc } = editor.state;

  const selectionLength = selection.to - selection.from;

  if (maxLength >= doc.content.size) {
    // If the document is smaller than the maximum length, return the entire document
    return {
      beforeSelection: doc.textBetween(
        0,
        selection.from,
        CONTEXT_BLOCK_SEPARATOR
      ),
      selection: doc.textBetween(
        selection.from,
        selection.to,
        CONTEXT_BLOCK_SEPARATOR
      ),
      afterSelection: doc.textBetween(
        selection.to,
        doc.content.size,
        CONTEXT_BLOCK_SEPARATOR
      ),
    };
  } else if (selectionLength > maxLength) {
    // If the selection is too large, truncate its middle to still allow continuations
    const selectionStart = doc.textBetween(
      selection.from,
      selection.from + Math.floor(maxLength / 2) - CONTEXT_TRUNCATION.length,
      CONTEXT_BLOCK_SEPARATOR
    );
    const selectionEnd = doc.textBetween(
      selection.to - Math.floor(maxLength / 2) + CONTEXT_TRUNCATION.length,
      selection.to,
      CONTEXT_BLOCK_SEPARATOR
    );

    return {
      beforeSelection: "",
      selection: `${selectionStart}${CONTEXT_TRUNCATION}${selectionEnd}`,
      afterSelection: "",
    };
  } else {
    // If the selection is smaller than (or equal to) the maximum length, extract as much as possible from the document around the selection

    // Start by taking as much as possible after the selection
    let beforeLength = Math.min(
      selection.from,
      Math.floor((maxLength - selectionLength) / 2)
    );
    const afterLength = Math.min(
      doc.content.size - selection.to,
      maxLength - selectionLength - beforeLength
    );

    // If needed (e.g. the selection is near the end), compensate before the selection
    if (beforeLength + afterLength + selectionLength < maxLength) {
      beforeLength = Math.min(
        selection.from,
        maxLength - selectionLength - afterLength
      );
    }

    let beforeSelection = doc.textBetween(
      Math.max(0, selection.from - beforeLength),
      selection.from,
      CONTEXT_BLOCK_SEPARATOR
    );
    let afterSelection = doc.textBetween(
      selection.to,
      Math.min(doc.content.size, selection.to + afterLength),
      CONTEXT_BLOCK_SEPARATOR
    );

    // Add leading truncation if `beforeSelection` doesn't contain the document's start
    if (selection.from - beforeLength > 0) {
      beforeSelection = `${CONTEXT_TRUNCATION}${beforeSelection}`;
    }

    // Add trailing truncation if `afterSelection` doesn't contain the document's end
    if (selection.to + afterLength < doc.content.size) {
      afterSelection = `${afterSelection}${CONTEXT_TRUNCATION}`;
    }

    return {
      beforeSelection,
      selection: doc.textBetween(
        selection.from,
        selection.to,
        CONTEXT_BLOCK_SEPARATOR
      ),
      afterSelection,
    };
  }
}
