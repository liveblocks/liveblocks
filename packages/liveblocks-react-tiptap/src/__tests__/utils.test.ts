import type { ContextualPromptContext } from "@liveblocks/core";
import { Editor } from "@tiptap/core";
import Document from "@tiptap/extension-document";
import Paragraph from "@tiptap/extension-paragraph";
import Text from "@tiptap/extension-text";

import { getContextualPromptContext } from "../utils";

function expectContextLength(
  context: ContextualPromptContext,
  expectedLength: number
) {
  // Allow for some variance in the total length because of textBetween's behavior and truncations
  const delta = 8;
  const length =
    context.beforeSelection.length +
    context.selection.length +
    context.afterSelection.length;

  expect(length).toBeGreaterThanOrEqual(expectedLength - delta);
  expect(length).toBeLessThanOrEqual(expectedLength + delta);
}

describe("utils", () => {
  describe("getContextualPromptContext", () => {
    let editor: Editor;

    beforeEach(() => {
      editor = new Editor({
        extensions: [Document, Text, Paragraph],
        // 2905 characters
        content:
          "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque viverra aliquet magna, id faucibus mauris porta non. Maecenas nec varius ipsum. Ut magna erat, venenatis eu neque nec, bibendum blandit lorem. Aliquam neque risus, tempor sed volutpat eget, cursus at magna. Nullam eu metus orci. Aliquam bibendum pellentesque elit at feugiat. Sed elementum nisl eget ante commodo, pellentesque consectetur mi semper. Vivamus pharetra justo sed velit venenatis, sed lobortis lacus mollis. Mauris finibus dolor non turpis rutrum rhoncus a in lacus. Nunc tempor enim arcu, sit amet egestas urna imperdiet non. Fusce et eros tempus, eleifend leo nec, dictum augue. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Duis sit amet laoreet leo, ornare imperdiet massa. Cras iaculis ligula sed commodo suscipit. Donec sit amet sem ut diam ullamcorper pulvinar vel quis mauris. Sed accumsan finibus egestas. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Pellentesque viverra consectetur lectus vitae rhoncus. Maecenas hendrerit nulla id tristique ullamcorper. Suspendisse gravida dapibus ante volutpat convallis. Quisque enim lectus, lobortis vitae turpis ut, tempor semper lectus. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam in quam rhoncus, euismod leo nec, molestie risus. Phasellus eu urna sit amet massa ultrices semper. Integer id felis in lorem faucibus laoreet eu et nulla. Nulla tempor lobortis turpis, sed placerat nulla elementum eget. Vestibulum congue eleifend hendrerit. Curabitur maximus sit amet lectus ac dapibus. Suspendisse potenti. Pellentesque vel est molestie, posuere orci nec, finibus ligula. Aliquam erat volutpat. Proin justo diam, lobortis a imperdiet non, mollis quis massa. Mauris rhoncus scelerisque leo, vulputate tempor sem commodo id. Donec volutpat ante at massa mattis ultricies. Mauris quis egestas felis, non vulputate dolor. Maecenas egestas diam at arcu auctor sagittis. Nulla dignissim feugiat erat. Sed justo enim, tincidunt quis nunc vel, vestibulum euismod libero. Suspendisse dictum at lorem sit amet porta. Vestibulum vehicula mattis leo sit amet pulvinar. Maecenas in hendrerit magna. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras dignissim porttitor quam nec interdum. Sed posuere leo eget sodales dignissim. Mauris malesuada vel diam eget faucibus. Proin et hendrerit dui, ac porttitor nibh. Praesent efficitur mollis condimentum. Aliquam sit amet turpis in felis vulputate eleifend ut facilisis arcu. Praesent feugiat tincidunt metus quis pulvinar. Suspendisse augue odio, finibus a interdum a, porta et ipsum. Praesent sed orci eu quam auctor scelerisque. Nunc lacinia congue aliquam. Donec odio nisl, posuere et elementum luctus, commodo volutpat diam. Praesent luctus id lacus eu cursus. Sed nec dolor quis neque convallis cursus et quis diam.",
      });
    });

    afterEach(() => {
      editor.destroy();
    });

    test("should handle an empty selection", () => {
      // Set the selection to a single point in the middle of the document
      editor.commands.setTextSelection({
        from: editor.state.doc.content.size / 2,
        to: editor.state.doc.content.size / 2,
      });

      const context = getContextualPromptContext(editor, 1000);

      expectContextLength(context, 1000);
      expect(context).toMatchSnapshot();
    });

    test("should prioritize the selection", () => {
      // Select 200 characters in the middle of the document
      editor.commands.setTextSelection({
        from: editor.state.doc.content.size / 2 - 100,
        to: editor.state.doc.content.size / 2 + 100,
      });

      const selectionText = editor.state.doc.textBetween(
        editor.state.selection.from,
        editor.state.selection.to,
        " "
      );
      const context = getContextualPromptContext(editor, 200);

      expectContextLength(context, 200);
      expect(context.selection.includes(selectionText)).toBeTruthy();
    });

    test("should truncate the selection if it is too large", () => {
      // Select the entire document
      editor.commands.setTextSelection({
        from: 0,
        to: editor.state.doc.content.size,
      });

      const context = getContextualPromptContext(editor, 1000);

      expectContextLength(context, 1000);
      expect(context).toMatchSnapshot();
    });

    test("should take into account the document size", () => {
      // Select from the 100th character to the 50th character from the end
      editor.commands.setTextSelection({
        from: 100,
        to: editor.state.doc.content.size - 50,
      });

      const context = getContextualPromptContext(editor, 10000);

      expectContextLength(context, editor.state.doc.content.size);
      expect(context).toMatchSnapshot();
    });

    test("should balance the available length equally on both sides of the selection", () => {
      // Select 200 characters in the middle of the document
      editor.commands.setTextSelection({
        from: editor.state.doc.content.size / 2 - 100,
        to: editor.state.doc.content.size / 2 + 100,
      });

      const context = getContextualPromptContext(editor, 1000);

      expectContextLength(context, 1000);
      expect(context).toMatchSnapshot();
    });

    test("should compensate after the selection if before is not enough", () => {
      // Select the first 200 characters
      editor.commands.setTextSelection({
        from: 0,
        to: 200,
      });

      const context = getContextualPromptContext(editor, 1000);

      expectContextLength(context, 1000);
      expect(context).toMatchSnapshot();
    });

    test("should compensate before the selection if after is not enough", () => {
      // Select 200 characters near the end of the document
      editor.commands.setTextSelection({
        from: editor.state.doc.content.size - 350,
        to: editor.state.doc.content.size - 150,
      });

      const context = getContextualPromptContext(editor, 1000);

      expectContextLength(context, 1000);
      expect(context).toMatchSnapshot();
    });

    test("should support elements", () => {
      editor.commands.clearContent();
      editor.commands.insertContent(`<p>${"lorem ".repeat(200)}</p>`);
      editor.commands.insertContent(
        "<ul><li>lorem ipsum</li><li><p>lorem ipsum</p></li><li><em>?</em></li></ul>"
      );
      editor.commands.insertContent(`<p>${"ipsum ".repeat(100)}</p>`);
      editor.commands.insertContent("<pre><code>const a = 1;</code></pre>");

      // Select 200 characters in the middle of the document
      editor.commands.setTextSelection({
        from: editor.state.doc.content.size / 2 - 100,
        to: editor.state.doc.content.size / 2 + 100,
      });

      const context = getContextualPromptContext(editor, 1000);

      expectContextLength(context, 1000);
      expect(context).toMatchSnapshot();
    });
  });
});
