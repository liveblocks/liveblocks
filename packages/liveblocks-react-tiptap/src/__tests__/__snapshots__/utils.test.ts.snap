// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getContextualPromptContext should balance the available length equally on both sides of the selection 1`] = `
{
  "afterSelection": "congue eleifend hendrerit. Curabitur maximus sit amet lectus ac dapibus. Suspendisse potenti. Pellentesque vel est molestie, posuere orci nec, finibus ligula. Aliquam erat volutpat. Proin justo diam, lobortis a imperdiet non, mollis quis massa. Mauris rhoncus scelerisque leo, vulputate tempor sem commodo id. Donec volutpat ante at massa mattis ultricies. Mauris quis egestas felis, non vulputate do[…]",
  "beforeSelection": "[…]ius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Pellentesque viverra consectetur lectus vitae rhoncus. Maecenas hendrerit nulla id tristique ullamcorper. Suspendisse gravida dapibus ante volutpat convallis. Quisque enim lectus, lobortis vitae turpis ut, tempor semper lectus. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam in quam rhoncus, euismod leo nec",
  "selection": ", molestie risus. Phasellus eu urna sit amet massa ultrices semper. Integer id felis in lorem faucibus laoreet eu et nulla. Nulla tempor lobortis turpis, sed placerat nulla elementum eget. Vestibulum ",
}
`;

exports[`utils getContextualPromptContext should compensate after the selection if before is not enough 1`] = `
{
  "afterSelection": "landit lorem. Aliquam neque risus, tempor sed volutpat eget, cursus at magna. Nullam eu metus orci. Aliquam bibendum pellentesque elit at feugiat. Sed elementum nisl eget ante commodo, pellentesque consectetur mi semper. Vivamus pharetra justo sed velit venenatis, sed lobortis lacus mollis. Mauris finibus dolor non turpis rutrum rhoncus a in lacus. Nunc tempor enim arcu, sit amet egestas urna imperdiet non. Fusce et eros tempus, eleifend leo nec, dictum augue. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Duis sit amet laoreet leo, ornare imperdiet massa. Cras iaculis ligula sed commodo suscipit. Donec sit amet sem ut diam ullamcorper pulvinar vel quis mauris. Sed accumsan finibus egestas. Orci varius natoque penatibus et magnis dis parturient […]",
  "beforeSelection": "",
  "selection": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque viverra aliquet magna, id faucibus mauris porta non. Maecenas nec varius ipsum. Ut magna erat, venenatis eu neque nec, bibendum b",
}
`;

exports[`utils getContextualPromptContext should compensate before the selection if after is not enough 1`] = `
{
  "afterSelection": "nisl, posuere et elementum luctus, commodo volutpat diam. Praesent luctus id lacus eu cursus. Sed nec dolor quis neque convallis cursus et quis diam.",
  "beforeSelection": "[…]s. Mauris quis egestas felis, non vulputate dolor. Maecenas egestas diam at arcu auctor sagittis. Nulla dignissim feugiat erat. Sed justo enim, tincidunt quis nunc vel, vestibulum euismod libero. Suspendisse dictum at lorem sit amet porta. Vestibulum vehicula mattis leo sit amet pulvinar. Maecenas in hendrerit magna. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras dignissim porttitor quam nec interdum. Sed posuere leo eget sodales dignissim. Mauris malesuada vel diam eget faucibus. Proin et hendrerit dui, ac porttitor nibh. Praesent efficitur mollis condimentum. Aliquam sit amet turpis in felis vulputate eleifend ut facilisis ar",
  "selection": "cu. Praesent feugiat tincidunt metus quis pulvinar. Suspendisse augue odio, finibus a interdum a, porta et ipsum. Praesent sed orci eu quam auctor scelerisque. Nunc lacinia congue aliquam. Donec odio ",
}
`;

exports[`utils getContextualPromptContext should handle an empty selection 1`] = `
{
  "afterSelection": "us laoreet eu et nulla. Nulla tempor lobortis turpis, sed placerat nulla elementum eget. Vestibulum congue eleifend hendrerit. Curabitur maximus sit amet lectus ac dapibus. Suspendisse potenti. Pellentesque vel est molestie, posuere orci nec, finibus ligula. Aliquam erat volutpat. Proin justo diam, lobortis a imperdiet non, mollis quis massa. Mauris rhoncus scelerisque leo, vulputate tempor sem commodo id. Donec volutpat ante at massa mattis ultricies. Mauris quis egestas felis, non vulputate do[…]",
  "beforeSelection": "[…]ius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Pellentesque viverra consectetur lectus vitae rhoncus. Maecenas hendrerit nulla id tristique ullamcorper. Suspendisse gravida dapibus ante volutpat convallis. Quisque enim lectus, lobortis vitae turpis ut, tempor semper lectus. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam in quam rhoncus, euismod leo nec, molestie risus. Phasellus eu urna sit amet massa ultrices semper. Integer id felis in lorem faucib",
  "selection": "",
}
`;

exports[`utils getContextualPromptContext should support elements 1`] = `
{
  "afterSelection": "rem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem 
lorem ipsum
lorem ipsum
?
ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum ipsum[…]",
  "beforeSelection": "[…]rem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem ",
  "selection": "lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lorem lo",
}
`;

exports[`utils getContextualPromptContext should take into account the document size 1`] = `
{
  "afterSelection": "c dolor quis neque convallis cursus et quis diam.",
  "beforeSelection": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque viverra aliquet magna, id fau",
  "selection": "cibus mauris porta non. Maecenas nec varius ipsum. Ut magna erat, venenatis eu neque nec, bibendum blandit lorem. Aliquam neque risus, tempor sed volutpat eget, cursus at magna. Nullam eu metus orci. Aliquam bibendum pellentesque elit at feugiat. Sed elementum nisl eget ante commodo, pellentesque consectetur mi semper. Vivamus pharetra justo sed velit venenatis, sed lobortis lacus mollis. Mauris finibus dolor non turpis rutrum rhoncus a in lacus. Nunc tempor enim arcu, sit amet egestas urna imperdiet non. Fusce et eros tempus, eleifend leo nec, dictum augue. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Duis sit amet laoreet leo, ornare imperdiet massa. Cras iaculis ligula sed commodo suscipit. Donec sit amet sem ut diam ullamcorper pulvinar vel quis mauris. Sed accumsan finibus egestas. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Pellentesque viverra consectetur lectus vitae rhoncus. Maecenas hendrerit nulla id tristique ullamcorper. Suspendisse gravida dapibus ante volutpat convallis. Quisque enim lectus, lobortis vitae turpis ut, tempor semper lectus. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam in quam rhoncus, euismod leo nec, molestie risus. Phasellus eu urna sit amet massa ultrices semper. Integer id felis in lorem faucibus laoreet eu et nulla. Nulla tempor lobortis turpis, sed placerat nulla elementum eget. Vestibulum congue eleifend hendrerit. Curabitur maximus sit amet lectus ac dapibus. Suspendisse potenti. Pellentesque vel est molestie, posuere orci nec, finibus ligula. Aliquam erat volutpat. Proin justo diam, lobortis a imperdiet non, mollis quis massa. Mauris rhoncus scelerisque leo, vulputate tempor sem commodo id. Donec volutpat ante at massa mattis ultricies. Mauris quis egestas felis, non vulputate dolor. Maecenas egestas diam at arcu auctor sagittis. Nulla dignissim feugiat erat. Sed justo enim, tincidunt quis nunc vel, vestibulum euismod libero. Suspendisse dictum at lorem sit amet porta. Vestibulum vehicula mattis leo sit amet pulvinar. Maecenas in hendrerit magna. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Cras dignissim porttitor quam nec interdum. Sed posuere leo eget sodales dignissim. Mauris malesuada vel diam eget faucibus. Proin et hendrerit dui, ac porttitor nibh. Praesent efficitur mollis condimentum. Aliquam sit amet turpis in felis vulputate eleifend ut facilisis arcu. Praesent feugiat tincidunt metus quis pulvinar. Suspendisse augue odio, finibus a interdum a, porta et ipsum. Praesent sed orci eu quam auctor scelerisque. Nunc lacinia congue aliquam. Donec odio nisl, posuere et elementum luctus, commodo volutpat diam. Praesent luctus id lacus eu cursus. Sed ne",
}
`;

exports[`utils getContextualPromptContext should truncate the selection if it is too large 1`] = `
{
  "afterSelection": "",
  "beforeSelection": "",
  "selection": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque viverra aliquet magna, id faucibus mauris porta non. Maecenas nec varius ipsum. Ut magna erat, venenatis eu neque nec, bibendum blandit lorem. Aliquam neque risus, tempor sed volutpat eget, cursus at magna. Nullam eu metus orci. Aliquam bibendum pellentesque elit at feugiat. Sed elementum nisl eget ante commodo, pellentesque consectetur mi semper. Vivamus pharetra justo sed velit venenatis, sed lobortis lacus mollis. Mauris[…]roin et hendrerit dui, ac porttitor nibh. Praesent efficitur mollis condimentum. Aliquam sit amet turpis in felis vulputate eleifend ut facilisis arcu. Praesent feugiat tincidunt metus quis pulvinar. Suspendisse augue odio, finibus a interdum a, porta et ipsum. Praesent sed orci eu quam auctor scelerisque. Nunc lacinia congue aliquam. Donec odio nisl, posuere et elementum luctus, commodo volutpat diam. Praesent luctus id lacus eu cursus. Sed nec dolor quis neque convallis cursus et quis diam.",
}
`;
